Src/ILGPU.Algorithms/MatrixOperations/MaskedSparseMatrixExtensions.cs

```diff
@@ -0,0 +1,390 @@
+    /// <param name="stream">The current accelerator stream.</param>
+    /// <param name="maskPredicate">
+    /// The input masking predicate (targeting a dense matrix of shape MxK).
+    /// </param>
+    /// <param name="aView">A dense input matrix of shape MxK.</param>
+    /// <param name="bView">A sparse matrix B of shape NxK (will transpose).</param>
+    /// <param name="outView">
+    /// A dense output view containing the results of the multiplication.
+    /// </param>
+    public delegate void MaskedSparseMatrixMultiplier<T, TPredicate, TStride>(
+        AcceleratorStream stream,
+        TPredicate maskPredicate,
+        ArrayView2D<T, TStride> aView,
+        SparseMatrixView<T, TStride> bView,
+        ArrayView2D<T, TStride> outView)
+        where T : unmanaged
+        where TStride : struct, IStride2D
+        where TPredicate : struct, InlineList.IPredicate<Index2D>;
+
+    /// <summary>
+    /// Specialized extensions to operate on sparse matrices while taking predicates
+    /// into account to skip specific elements.
+    /// </summary>
+    public static class MaskedSparseMatrixExtensions
+    {
+        /// <summary>
+        /// The matrix multiplication kernel that runs on the accelerated device while
+        /// using thread compaction to free masked warps.
+        /// </summary>
+        /// <param name="maskPredicate">
+        /// The input masking predicate (targeting a dense matrix of shape MxK).
+        /// </param>
+        /// <param name="aView">A dense input matrix of shape MxK.</param>
+        /// <param name="bView">A sparse matrix B of shape NxK (will transpose).</param>
+        /// <param name="outView">
+        /// A dense output view containing the results of the multiplication.
+        /// </param>
+        /// <param name="processor">An instance of an actual mul-add operation.</param>
+        internal static void MaskedSparseMatrixMultiplierKernel<
+            T,
+            TPredicate,
+            TStride,
+            TProcessor>(
+            TPredicate maskPredicate,
+            ArrayView2D<T, TStride> aView,
+            SparseMatrixView<T, TStride> bView,
+            ArrayView2D<T, TStride> outView,
+            TProcessor processor)
+            where T : unmanaged
+            where TStride : struct, IStride2D
+            where TPredicate : struct, InlineList.IPredicate<Index2D>
+            where TProcessor : struct, IMaskedSparseMatrixProcessor<T>
+        {
+            var maskView = SharedMemory.GetDynamic<int>();
+            Trace.Assert(
+                Group.DimX <= maskView.IntLength,
+                "Invalid shared memory config");
+
+            // Get all predicate results and perform thread compaction
+            int index = Grid.GlobalLinearIndex;
+            var maskIndex2D = aView.Stride.ReconstructFromElementIndex(index);
+            bool apply = maskIndex2D.X < aView.Extent.X & maskIndex2D.Y < aView.Extent.Y;
```

Comment:
What happens when apply is false here?  That is, the predicate is out of range?  Shouldn't we throw an error?

Reply:
It can happen that the index will be out of range based on the current grid/group configuration that we dispatch. Threads processing oor elements will just be skipped and terminate peacefully.